README.md

```diff
@@ -0,0 +1,112 @@
+# Secure AWS Instance Profiles on DC/OS
+Security is always an important topic in today's distributed systems. With DC/OS Enterprise we offer a feature called DC/OS Secrets. With Secrets it is possible to inject secure information like passwords or cryptographic keys into your application. No other Application is able to read or change this information and with DC/OS IAM you can also strip down the group of users having access to this information.
+
+## The usual workflow
+Let's assume you have an application that wants to access AWS resources, like a S3 bucket. With Secrets you can easily create an IAM user and assign it a Policy which is able to access your particular bucket. You take the ACCESS_KEY and add it to your Marathon specification and store the SECRET_ACCESS_KEY into a DC/OS Secret in your default vault which you also specify in your marathon application. This is absolutely not bad but it means you have to rotate these credentials on a regular basis and therefore you need to update your application from time to time.
+
+## AWS Agents can do better
+If your agents are already running on AWS Instances there is a way better and best practise solution solving your problem: Instance Profiles. Instance profiles allow you to assign Roles to Instances. The AWS SDK will realise running on an AWS Instance and try to retrieve credentials from the AWS Metadata API. The huge benefit of this is you do not need to rotate credentials by yourself as AWS is taking care of it. These credentials will have a short live time so even if they get leaked a user will only have a certain amount of time to use those credentials.
+
+## Not every task should have this privilege
+In DC/OS multiple applications will share the same agent and therefore share the same instance profile. This is something you might want to avoid. With the initially described process you only hand out the users credentials to the applications you’ve selected, so you decide based on the secret containing the credentials which application gets the creds.
+
+## AssumeRole and external_id
+We can combine the security of instance profiles with the selective authorization of DC/OS secrets. AWS offers a process called AssumeRole. With this a Role (Instance,User) is able to retrieve temporary credentials for another role ( even other accounts are possible ).
+So in our example the Instance would Assume into the Role having access to the S3 bucket.
+This process alone does not really change the authorization problem as still every application would be able to use it but AWS gives an additional layer of security to this procedure: external_id. The external_id is a pre-shared-key added to the trust relationship of a role which allows us to assume into it.
+This PSK will allow us to use DC/OS Secrets acting as authorization instance for our application by placing an AssumeRole configuration including the external_id.
```

Comment:
We can combine the security of instance profiles with the selective authorization of DC/OS secrets. AWS offers a process called AssumeRole. With this a Role (Instance,User) is able to retrieve temporary credentials for another role (even in other AWS accounts).
So in our example, the Instance would Assume a Role that has access to the S3 bucket.
This process alone does not really change the authorization problem as every application would still be able to use it but AWS gives an additional layer of security to this procedure called: external_id. The external_id is a pre-shared-key added to the trust relationship of a role which allows us to assume this role.
This PSK will allow us to use DC/OS Secrets acting as an authorization instance for our application by placing an AssumeRole configuration that includes the external_id.